### ISSUE TITLE
NuGetFallbackFolder resolving File Version conflicts for same Package Versions

### ISSUE THREAD

**Issue by MattLavalleeMA** [dotnet/sdk#2780]
I have a .NET Core 2.2 solution (migrated from 2.1) with eight projects: two AspNetCore projects, five class libraries, and a unit test project.  Recently, I started receiving version resolution "messages" for _HandlePackageFileConflicts like this:

> C:\Program Files\dotnet\sdk\2.2.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:C:\Users\mlavallee\.nuget\packages\system.security.principal.windows\4.5.1\lib\netstandard2.0\System.Security.Principal.Windows.dll'.  NETSDK1034: Choosing 'Platform:System.Security.Principal.Windows.dll' because file version '4.6.27110.4' is greater than '4.6.26919.2'.

and

> C:\Program Files\dotnet\sdk\2.2.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.AccessControl.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.accesscontrol\4.5.0\lib\netstandard2.0\System.Security.AccessControl.dll'.  NETSDK1034: Choosing 'Platform:System.Security.AccessControl.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.

We've all had to deal with binding _package version_ conflicts before, but in this case it seems to be at the _file version_ level, and with references that are included/embedded in the .NET Core 2.2 SDK itself.  Looking at the manifest file:
`C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.2.0\build\netcoreapp2.2\Microsoft.NETCore.App.PlatformManifest.txt`
>System.Security.Principal.Windows.dll|Microsoft.NETCore.App|4.1.1.1|4.6.27110.4
>System.Security.AccessControl.dll|Microsoft.NETCore.App|4.1.1.0|4.6.27110.4

However, the newest packages on NuGet have the file versions noted in the conflicts above and are from much older build sources (9/18/2018 vs. 12/6/2018).  So, a few questions from this:

1. Why is the SDK embedded reference file newer than what's available on NuGet?
2. Why does the NuGetFallbackFolder have earlier versions than the SDK?
3. Why are the file versions for the same package versions different?
4. What's the purpose of the NuGetFallbackFolder if I also have a package cache? (This seems like pure bloat)

Although these aren't raised as build warnings, just _messages_, it seems like I can't have confidence in the reference code being consistent between builds, so either they should be warnings or there should be a way of resolving them.

Thanks,
Matt

**Comment by jinweijie** [comment]
I encountered same issue:

`Target _HandlePackageFileConflicts:
		Encountered conflict between 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Collections.Immutable.dll' and 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.collections.immutable/1.5.0/lib/netstandard2.0/System.Collections.Immutable.dll'.  NETSDK1034: Choosing 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Collections.Immutable.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Diagnostics.DiagnosticSource.dll' and 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.diagnostics.diagnosticsource/4.5.0/lib/netstandard1.3/System.Diagnostics.DiagnosticSource.dll'.  NETSDK1034: Choosing 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Diagnostics.DiagnosticSource.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Reflection.Metadata.dll' and 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.reflection.metadata/1.6.0/lib/netstandard2.0/System.Reflection.Metadata.dll'.  NETSDK1034: Choosing 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Reflection.Metadata.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Runtime.Serialization.Xml.dll' and 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.runtime.serialization.xml/4.3.0/ref/netstandard1.3/System.Runtime.Serialization.Xml.dll'.  NETSDK1033: Choosing 'Reference:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.netcore.app/2.2.0/ref/netcoreapp2.2/System.Runtime.Serialization.Xml.dll' because AssemblyVersion '4.1.4.0' is greater than '4.1.1.0'.
		Encountered conflict between 'Platform:Microsoft.Win32.Registry.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.win32.registry/4.5.0/lib/netstandard2.0/Microsoft.Win32.Registry.dll'.  NETSDK1034: Choosing 'Platform:Microsoft.Win32.Registry.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Collections.Immutable.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.collections.immutable/1.5.0/lib/netstandard2.0/System.Collections.Immutable.dll'.  NETSDK1034: Choosing 'Platform:System.Collections.Immutable.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Diagnostics.DiagnosticSource.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.diagnostics.diagnosticsource/4.5.0/lib/netstandard1.3/System.Diagnostics.DiagnosticSource.dll'.  NETSDK1033: Choosing 'Platform:System.Diagnostics.DiagnosticSource.dll' because AssemblyVersion '4.0.3.1' is greater than '4.0.3.0'.
		Encountered conflict between 'Platform:System.Reflection.Metadata.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.reflection.metadata/1.6.0/lib/netstandard2.0/System.Reflection.Metadata.dll'.  NETSDK1034: Choosing 'Platform:System.Reflection.Metadata.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Runtime.Serialization.Xml.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.runtime.serialization.xml/4.3.0/lib/netstandard1.3/System.Runtime.Serialization.Xml.dll'.  NETSDK1033: Choosing 'Platform:System.Runtime.Serialization.Xml.dll' because AssemblyVersion '4.1.4.0' is greater than '4.1.2.0'.
		Encountered conflict between 'Platform:System.Security.AccessControl.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.accesscontrol/4.5.0/lib/netstandard2.0/System.Security.AccessControl.dll'.  NETSDK1034: Choosing 'Platform:System.Security.AccessControl.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Security.Cryptography.Cng.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.cryptography.cng/4.5.0/lib/netcoreapp2.1/System.Security.Cryptography.Cng.dll'.  NETSDK1034: Choosing 'Platform:System.Security.Cryptography.Cng.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.principal.windows/4.5.0/lib/netstandard2.0/System.Security.Principal.Windows.dll'.  NETSDK1033: Choosing 'Platform:System.Security.Principal.Windows.dll' because AssemblyVersion '4.1.1.1' is greater than '4.1.1.0'.
		Encountered conflict between 'Platform:Microsoft.Win32.Registry.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.win32.registry/4.5.0/runtimes/unix/lib/netstandard2.0/Microsoft.Win32.Registry.dll'.  NETSDK1034: Choosing 'Platform:Microsoft.Win32.Registry.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:Microsoft.Win32.Registry.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/microsoft.win32.registry/4.5.0/runtimes/win/lib/netstandard2.0/Microsoft.Win32.Registry.dll'.  NETSDK1034: Choosing 'Platform:Microsoft.Win32.Registry.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Security.AccessControl.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.accesscontrol/4.5.0/runtimes/win/lib/netcoreapp2.0/System.Security.AccessControl.dll'.  NETSDK1034: Choosing 'Platform:System.Security.AccessControl.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Security.Cryptography.Cng.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.cryptography.cng/4.5.0/runtimes/win/lib/netcoreapp2.1/System.Security.Cryptography.Cng.dll'.  NETSDK1034: Choosing 'Platform:System.Security.Cryptography.Cng.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
		Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.principal.windows/4.5.0/runtimes/unix/lib/netcoreapp2.0/System.Security.Principal.Windows.dll'.  NETSDK1033: Choosing 'Platform:System.Security.Principal.Windows.dll' because AssemblyVersion '4.1.1.1' is greater than '4.1.1.0'.
		Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:/usr/local/share/dotnet/sdk/NuGetFallbackFolder/system.security.principal.windows/4.5.0/runtimes/win/lib/netcoreapp2.0/System.Security.Principal.Windows.dll'.  NETSDK1033: Choosing 'Platform:System.Security.Principal.Windows.dll' because AssemblyVersion '4.1.1.1' is greater than '4.1.1.0'.
	`

**Comment by kcliffe** [comment]
Me also. dotnet 2.1 upgraded to 2.2. 

```
C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Reference:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.2.0\ref\netcoreapp2.2\System.Collections.Immutable.dll' and 'Reference:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.collections.immutable\1.5.0\lib\netstandard2.0\System.Collections.Immutable.dll'.  NETSDK1034: Choosing 'Reference:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.2.0\ref\netcoreapp2.2\System.Collections.Immutable.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Reference:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.2.0\ref\netcoreapp2.2\System.Diagnostics.DiagnosticSource.dll' and 'Reference:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.diagnostics.diagnosticsource\4.5.0\lib\netstandard1.3\System.Diagnostics.DiagnosticSource.dll'.  NETSDK1034: Choosing 'Reference:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.2.0\ref\netcoreapp2.2\System.Diagnostics.DiagnosticSource.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:Microsoft.Win32.Registry.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.win32.registry\4.5.0\lib\netstandard2.0\Microsoft.Win32.Registry.dll'.  NETSDK1034: Choosing 'Platform:Microsoft.Win32.Registry.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Collections.Immutable.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.collections.immutable\1.5.0\lib\netstandard2.0\System.Collections.Immutable.dll'.  NETSDK1034: Choosing 'Platform:System.Collections.Immutable.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Diagnostics.DiagnosticSource.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.diagnostics.diagnosticsource\4.5.0\lib\netstandard1.3\System.Diagnostics.DiagnosticSource.dll'.  NETSDK1033: Choosing 'Platform:System.Diagnostics.DiagnosticSource.dll' because AssemblyVersion '4.0.3.1' is greater than '4.0.3.0'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Reflection.DispatchProxy.dll' and 'Runtime:C:\Users\kwc\.nuget\packages\system.reflection.dispatchproxy\4.5.0\lib\netcoreapp2.0\System.Reflection.DispatchProxy.dll'.  NETSDK1034: Choosing 'Platform:System.Reflection.DispatchProxy.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.AccessControl.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.accesscontrol\4.5.0\lib\netstandard2.0\System.Security.AccessControl.dll'.  NETSDK1034: Choosing 'Platform:System.Security.AccessControl.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.Cryptography.Cng.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.cryptography.cng\4.5.0\lib\netcoreapp2.1\System.Security.Cryptography.Cng.dll'.  NETSDK1034: Choosing 'Platform:System.Security.Cryptography.Cng.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.principal.windows\4.5.0\lib\netstandard2.0\System.Security.Principal.Windows.dll'.  NETSDK1033: Choosing 'Platform:System.Security.Principal.Windows.dll' because AssemblyVersion '4.1.1.1' is greater than '4.1.1.0'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:Microsoft.Win32.Registry.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.win32.registry\4.5.0\runtimes\unix\lib\netstandard2.0\Microsoft.Win32.Registry.dll'.  NETSDK1034: Choosing 'Platform:Microsoft.Win32.Registry.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:Microsoft.Win32.Registry.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.win32.registry\4.5.0\runtimes\win\lib\netstandard2.0\Microsoft.Win32.Registry.dll'.  NETSDK1034: Choosing 'Platform:Microsoft.Win32.Registry.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.AccessControl.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.accesscontrol\4.5.0\runtimes\win\lib\netcoreapp2.0\System.Security.AccessControl.dll'.  NETSDK1034: Choosing 'Platform:System.Security.AccessControl.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.Cryptography.Cng.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.cryptography.cng\4.5.0\runtimes\win\lib\netcoreapp2.1\System.Security.Cryptography.Cng.dll'.  NETSDK1034: Choosing 'Platform:System.Security.Cryptography.Cng.dll' because file version '4.6.27110.4' is greater than '4.6.26515.6'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.principal.windows\4.5.0\runtimes\unix\lib\netcoreapp2.0\System.Security.Principal.Windows.dll'.  NETSDK1033: Choosing 'Platform:System.Security.Principal.Windows.dll' because AssemblyVersion '4.1.1.1' is greater than '4.1.1.0'.
1>    C:\Program Files\dotnet\sdk\2.2.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(41,5): message NETSDK1041: Encountered conflict between 'Platform:System.Security.Principal.Windows.dll' and 'Runtime:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.security.principal.windows\4.5.0\runtimes\win\lib\netcoreapp2.0\System.Security.Principal.Windows.dll'.  NETSDK1033: Choosing 'Platform:System.Security.Principal.Windows.dll' because AssemblyVersion '4.1.1.1' is greater than '4.1.1.0'.

```
Happens on local box and Jenkins CI box.

Not even sure if we should be nervous here that one day our builds just stop successfully building because something that's maybe randomly choosing the correct DLL sudfenly stops doing that...

**Comment by dsplaisted** [comment]
@nguerrera Was this fixed with #3327?

**Comment by nguerrera** [comment]
Yes, fixed by #3327
